# <center> Titanic Dataset</center>

In [129]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OneHotEncoder

from sklearn.metrics import accuracy_score

In [2]:
train_data = pd.read_csv('titanic_data/train.csv')
test_data = pd.read_csv('titanic_data/test.csv')
gender = pd.read_csv('titanic_data/gender_submission.csv')

In [3]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [7]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [8]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


First we have to handle missing values then convert categorical values to numerical value. <br>
To handle both dataframe together we can concatenate and train and test dataframe.

In [11]:
# First insert survived data to test data 
test_data.insert(1, value=np.NaN, column='Survived')

In [12]:
test_data

PassengerId  Survived  Pclass  \
0            892       NaN       3   
1            893       NaN       3   
2            894       NaN       2   
3            895       NaN       3   
4            896       NaN       3   
..           ...       ...     ...   
413         1305       NaN       3   
414         1306       NaN       1   
415         1307       NaN       3   
416         1308       NaN       3   
417         1309       NaN       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

In [15]:
# Now test and train data have same length. Therefore we can concatenate them
df = pd.concat([train_data, test_data], axis=0, ignore_index=True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  1309 non-null   int64  
 1   Survived     891 non-null    float64
 2   Pclass       1309 non-null   int64  
 3   Name         1309 non-null   object 
 4   Sex          1309 non-null   object 
 5   Age          1046 non-null   float64
 6   SibSp        1309 non-null   int64  
 7   Parch        1309 non-null   int64  
 8   Ticket       1309 non-null   object 
 9   Fare         1308 non-null   float64
 10  Cabin        295 non-null    object 
 11  Embarked     1307 non-null   object 
dtypes: float64(3), int64(4), object(5)
memory usage: 122.8+ KB


### Handling Missing Values

Survived columns is empty for test case and we will predict with ML. For other attributes(age, fare, embarked and cabin) we have to handle missing data. 

#### 1. Cabin Column

In [17]:
# As 'Cabin' column has ~78% nan value, we can just remove it from dataset
df.drop(columns='Cabin', inplace=True)

#### 2. Embarked

In [18]:
df[df.Embarked.isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62       1.0       1                        Icard, Miss. Amelie   
829          830       1.0       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Embarked  
61   female  38.0      0      0  113572  80.0      NaN  
829  female  62.0      0      0  113572  80.0      NaN

We have two woman passenger and both paid $80 fare, first class. 
And both survived.

In [48]:
df[(df.Pclass == 1) & (df.Fare < 80) & (df.Fare > 70) & (df.Sex == 'female')]['Embarked'].value_counts(normalize=True)

C    0.529412
S    0.470588
Name: Embarked, dtype: float64

When we control this conditions we see that they are not from Q station. And the chance of embarked from C or S is very close to each other. <br>
Therefore I will assign C to one and S to the other. <br>
Note: But in reality from their name when we search we know that both embarked from S.

In [50]:
df.loc[61, 'Embarked'] = 'S'
df.loc[829, 'Embarked'] = 'C'

#### 3. Fare

In [52]:
df[df.Fare.isnull()]

PassengerId  Survived  Pclass                Name   Sex   Age  SibSp  \
1043         1044       NaN       3  Storey, Mr. Thomas  male  60.5      0   

      Parch Ticket  Fare Embarked  
1043      0   3701   NaN        S

Fare according to Class and Station

In [64]:
df.groupby(by=['Embarked', 'Pclass'])['Fare'].mean()

Embarked  Pclass
C         1         106.656279
          2          23.300593
          3          11.021624
Q         1          90.000000
          2          11.735114
          3          10.390820
S         1          72.192206
          2          21.206921
          3          14.435422
Name: Fare, dtype: float64

In [67]:
df.loc[1043, 'Fare'] = df.groupby(by=['Embarked', 'Pclass'])['Fare'].mean().loc[('S',3)]

### 4. Age

In [73]:
df[df.Age.isnull()]

PassengerId  Survived  Pclass                             Name     Sex  \
5               6       0.0       3                 Moran, Mr. James    male   
17             18       1.0       2     Williams, Mr. Charles Eugene    male   
19             20       1.0       3          Masselmani, Mrs. Fatima  female   
26             27       0.0       3          Emir, Mr. Farred Chehab    male   
28             29       1.0       3    O'Dwyer, Miss. Ellen "Nellie"  female   
...           ...       ...     ...                              ...     ...   
1299         1300       NaN       3  Riordan, Miss. Johanna Hannah""  female   
1301         1302       NaN       3           Naughton, Miss. Hannah  female   
1304         1305       NaN       3               Spector, Mr. Woolf    male   
1307         1308       NaN       3              Ware, Mr. Frederick    male   
1308         1309       NaN       3         Peter, Master. Michael J    male   

      Age  SibSp  Parch     Ticket     Fare Embarked  
5     NaN      0      0     330877   8.4583        Q  
17    NaN      0      0     244373  13.0000        S  
19    NaN      0      0       2649   7.2250        C  
26    NaN      0      0       2631   7.2250        C  
28    NaN      0      0     330959   7.8792        Q  
...   ...    ...    ...        ...      ...      ...  
1299  NaN      0      0     334915   7.7208        Q  
1301  NaN      0      0     365237   7.7500        Q  
1304  NaN      0      0  A.5. 3236   8.0500        S  
1307  NaN      0      0     359309   8.0500        S  
1308  NaN      1      1       2668  22.3583        C  

[263 rows x 11 columns]

It seems we only have try to predict age information from class, sex and maybe ticket price. <br>
PS: We may use 'Name' columns for a better prediction.

In [81]:
df.Age = df.groupby(by=['Pclass','Sex'])['Age'].apply(lambda x: x.fillna(x.median()))

In [82]:
df.isnull().sum()

PassengerId      0
Survived       418
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         0
dtype: int64

Now we have no more missing values in dataframe

## ML Model

### Handling Categorical Columns

In [83]:
df

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  25.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  25.0      0   
1308                           Peter, Master. Michael J    male  25.0      1   

      Parch              Ticket      Fare Embarked  
0         0           A/5 21171    7.2500        S  
1         0            PC 17599   71.2833        C  
2         0    STON/O2. 3101282    7.9250        S  
3         0              113803   53.1000        S  
4         0              373450    8.0500        S  
...     ...                 ...       ...      ...  
1304      0           A.5. 3236    8.0500        S  
1305      0            PC 17758  108.9000        C  
1306      0  SOTON/O.Q. 3101262    7.2500        S  
1307      0              359309    8.0500        S  
1308      1                2668   22.3583        C  

[1309 rows x 11 columns]

We will drop columns we don't use in ML model like Name column, PassengerId, Ticket.
And also convert categorical data like 'Embarked', 'Sex' to numerical values.

In [86]:
df.drop(columns=['PassengerId', 'Name', 'Ticket'], inplace=True)

For embarked column we can use pandas get_dummies method or OneHotEncoder. Here I will use get_dummies. And concat with df

In [100]:
mt = df.copy()

In [104]:
df[['C','Q','S']] = pd.get_dummies(df['Embarked'])

Convert sex columns from categorical to numercial

In [108]:
df.Sex = df.Sex.map({'female':0, 'male':1})

Remove embarked column

In [114]:
df.drop(columns='Embarked', inplace=True)

### Using ML Model

In [120]:
train_final = df[df.Survived.notna()]

In [121]:
train_final

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  C  Q  S
0         0.0       3    1  22.0      1      0   7.2500  0  0  1
1         1.0       1    0  38.0      1      0  71.2833  1  0  0
2         1.0       3    0  26.0      0      0   7.9250  0  0  1
3         1.0       1    0  35.0      1      0  53.1000  0  0  1
4         0.0       3    1  35.0      0      0   8.0500  0  0  1
..        ...     ...  ...   ...    ...    ...      ... .. .. ..
886       0.0       2    1  27.0      0      0  13.0000  0  0  1
887       1.0       1    0  19.0      0      0  30.0000  0  0  1
888       0.0       3    0  22.0      1      2  23.4500  0  0  1
889       1.0       1    1  26.0      0      0  30.0000  1  0  0
890       0.0       3    1  32.0      0      0   7.7500  0  1  0

[891 rows x 10 columns]

In [122]:
test_final = df[df.Survived.isnull()]

In [123]:
test_final

Survived  Pclass  Sex   Age  SibSp  Parch      Fare  C  Q  S
891        NaN       3    1  34.5      0      0    7.8292  0  1  0
892        NaN       3    0  47.0      1      0    7.0000  0  0  1
893        NaN       2    1  62.0      0      0    9.6875  0  1  0
894        NaN       3    1  27.0      0      0    8.6625  0  0  1
895        NaN       3    0  22.0      1      1   12.2875  0  0  1
...        ...     ...  ...   ...    ...    ...       ... .. .. ..
1304       NaN       3    1  25.0      0      0    8.0500  0  0  1
1305       NaN       1    0  39.0      0      0  108.9000  1  0  0
1306       NaN       3    1  38.5      0      0    7.2500  0  0  1
1307       NaN       3    1  25.0      0      0    8.0500  0  0  1
1308       NaN       3    1  25.0      1      1   22.3583  1  0  0

[418 rows x 10 columns]

In [125]:
X_train = train_final.iloc[:,1:]
X_test = test_final.iloc[:,1:]
y_train = train_final.iloc[:,0]
y_test = gender['Survived']

In [126]:
forest = RandomForestClassifier(n_estimators=100, random_state=42)

In [127]:
forest.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [128]:
y_predict = forest.predict(X_test)

Predict result is:

In [131]:
print(f'Accuracy is {accuracy_score(y_test, y_predict)*100:.2f}%')

Accuracy is 81.58%


Our model accuracy is 81.58%. <br>
But it is obvious that the model can be improved. 